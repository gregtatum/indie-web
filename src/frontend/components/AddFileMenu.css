.addFileMenuError {
  margin: 8px;
  background: #ffbfbf;
  padding: 8px;
  border-radius: 3px;
  border: 1px solid #f00;
  font-size: var(--font-size-small);
}

.addFileMenuInput {
  flex: 1;
  font-size: var(--font-size-small);
}

.addFileMenuForm {
  display: flex;
  gap: 8px;
  margin: 8px;
  max-width: 300px;
}

.button.addFileMenuButton {
  margin: 8px;
}
