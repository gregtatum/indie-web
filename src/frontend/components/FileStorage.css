.file-storage-pre {
  background-color: #eee;
  border-radius: 4px;
  border: 2px solid #ddd;
  padding: 1rem;
}

.file-storage-input {
  display: flex;
  flex-direction: column;
  margin: 1rem 0;
  gap: 4px;

  input {
    padding: 0.5rem;
  }
}

.file-storage-list {
  border: 1px solid #ddd;
  background-color: #eee;
  border-radius: 4px;
  padding: 1rem;
  display: flex;
  flex-direction: column;
  gap: 10px;

  & > form {
    display: grid;
    grid-template-columns: 1fr 1fr auto;
    gap: 10px;
    margin: 0;
  }

  & input {
    padding: 0.5rem;
  }
}
