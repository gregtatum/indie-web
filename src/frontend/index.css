body {
  position: absolute;
  height: 100%;
  width: 100%;
  display: flex;
  margin: 0;
  justify-content: center;
  align-items: center;
  flex-direction: column;
  font-family: sans-serif;
}
