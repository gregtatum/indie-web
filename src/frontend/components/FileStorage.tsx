import * as React from 'react';

import './FileStorage.css';
import dedent from 'dedent';
import { $, $$, A, T, Hooks } from 'frontend';

export function FileStorage() {
  const fileStoreServers = $$.getFileStoreServers();
  const { dispatch, getState } = Hooks.useStore();
  const nameRef = React.useRef<null | HTMLInputElement>(null);
  const urlRef = React.useRef<null | HTMLInputElement>(null);
  const { error, setError } = Hooks.useError();

  function addFileServer(event: Event) {
    event.preventDefault();
    setError(null);
    const servers = $.getFileStoreServers(getState());
    const name = nameRef.current?.value;
    const url = urlRef.current?.value;
    const server = validateFileStoreServer(name, url, servers, setError);
    if (server) {
      dispatch(A.addFileStoreServer(server));
    }
  }

  return (
    <div className="page">
      <div className="pageInner">
        <h1>Manage File Storage</h1>
        <FileStorageList fileStoreServers={fileStoreServers} />
        <h2>Add a File Storage Server</h2>
        <p>
          Whether you are accessing files stored locally on your machine, or you
          want to access files on your NAS (Network Attached Storage), you can
          run the file storage server, and mount a folder on a machine you have
          access to.
        </p>

        <h3>1: Start the server</h3>
        <pre className="file-storage-pre">
          {dedent`
              git clone git@github.com:gregtatum/indie-web.git
              cd indie-web/src/server
              npm install
              npm start
          `}
        </pre>
        <p>Follow the installation instructions.</p>

        <h3>2: Add the Server</h3>
        <form onSubmit={(event) => addFileServer(event.nativeEvent)}>
          <div className="file-storage-input">
            <label htmlFor="file-storage-name">File Storage Name</label>
            <input
              type="text"
              id="file-storage-name"
              placeholder="NAS Storage"
              ref={nameRef}
              maxLength={200}
            ></input>
          </div>
          <div className="file-storage-input">
            <label htmlFor="file-storage-name">File Storage Address</label>
            <input
              type="text"
              id="file-storage-name"
              ref={urlRef}
              defaultValue="http://localhost:6543"
            ></input>
          </div>
          {error}
          <button type="submit">Add File Server</button>
        </form>
      </div>
    </div>
  );
}

interface FileStorageListProps {
  fileStoreServers: T.FileStoreServer[];
}

function FileStorageList({ fileStoreServers }: FileStorageListProps) {
  const { dispatch, getState } = Hooks.useStore();
  const { error, setError } = Hooks.useError();

  if (fileStoreServers.length === 0) {
    return null;
  }
  function update(event: Event, oldServer: T.FileStoreServer) {
    event.preventDefault();
    const { target } = event;
    if (!(target instanceof HTMLElement)) {
      return;
    }
    const form = target.closest('form');
    if (!form) {
      return;
    }
    const [nameInput, urlInput] = form.querySelectorAll('input');
    const otherServers = $.getFileStoreServers(getState()).filter(
      (server) =>
        server.name !== oldServer.name && server.url !== oldServer.url,
    );
    if (
      oldServer.name === nameInput.value &&
      oldServer.url === urlInput.value
    ) {
      // Nothing changed.
      return;
    }
    const server = validateFileStoreServer(
      nameInput.value,
      urlInput.value,
      otherServers,
      setError,
    );
    if (server) {
      dispatch(A.updateFileStoreServer(oldServer, server));
      dispatch(
        A.addMessage({
          message: <>File store server updated</>,
          timeout: true,
        }),
      );
    }
  }

  return (
    <>
      {error}
      <div className="file-storage-list">
        {fileStoreServers.map(({ name, url }) => (
          <form
            key={name + url}
            onSubmit={(event) => update(event.nativeEvent, { name, url })}
          >
            <input
              type="text"
              defaultValue={name}
              onBlur={(event) => update(event.nativeEvent, { name, url })}
            ></input>
            <input
              type="text"
              defaultValue={url}
              onBlur={(event) => update(event.nativeEvent, { name, url })}
            ></input>
            <input type="submit" style={{ display: 'none' }}></input>
            <button
              type="button"
              onClick={() => {
                if (
                  confirm(
                    'Are you sure you want to remove this file store server?',
                  )
                ) {
                  dispatch(A.removeFileStoreServer({ name, url }));
                }
              }}
            >
              Remove
            </button>
          </form>
        ))}
      </div>
    </>
  );
}

function validateFileStoreServer(
  name: string | undefined,
  url: string | undefined,
  servers: T.FileStoreServer[],
  setError: (error: string) => void,
): null | T.FileStoreServer {
  // Make sure all the values are set.
  if (!name) {
    setError('The name of the server must be set');
    return null;
  }
  if (!url) {
    setError('The address of the server must be set');
    return null;
  }

  // Validate the URL.
  try {
    new URL(url);
  } catch {
    url = 'http://' + url;
    try {
      new URL(url);
    } catch {
      setError('The address of the server was not a valid URL.');
      return null;
    }
  }

  if (name.length > 200) {
    setError('The server name should be less than 200 characters.');
    return null;
  }

  if (servers.some((server) => server.name === name)) {
    setError(`Another server already has the name "${name}"`);
    return null;
  }
  const existingURL = servers.find((server) => server.url === url);
  if (existingURL) {
    setError(`The server "${existingURL.name}" already uses that URL.`);
    return null;
  }

  return { name, url };
}
