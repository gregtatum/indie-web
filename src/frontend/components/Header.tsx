import * as React from 'react';
import * as Router from 'react-router-dom';
import { $$, A, Hooks } from 'frontend';
import { assertType, isAppSettingScrollTop } from 'frontend/utils';
import {
  getBrowserName,
  getFileStoreDisplayName,
} from 'frontend/logic/app-logic';
import './Header.css';
import { Menu, MenuButton } from './Menus';
import { overlayPortal } from 'frontend/hooks';

export function Header() {
  const view = $$.getView();
  const path = $$.getActiveFileDisplayPath();
  const [shouldHideHeader, setShouldHideHeader] = React.useState(false);
  const key = (view ?? '') + path;

  const headerStyle: React.CSSProperties = {};
  if (shouldHideHeader) {
    headerStyle.transform = 'translateY(var(--header-transform-y))';
  }

  React.useEffect(function trackScrolling() {
    const { scrollingElement } = document;
    if (!scrollingElement) {
      return () => {};
    }
    let headerPadding: number;
    {
      const headerPaddingStr =
        getComputedStyle(scrollingElement).getPropertyValue('--header-padding');
      if (!headerPaddingStr) {
        throw new Error('Expected to find a headerPadding style');
      }
      headerPaddingStr.replace('px', '');
      headerPadding = parseInt(headerPaddingStr, 10) * 0.5;
    }

    let _prevScroll = 0;
    let _shouldHideHeader = false;

    const onScroll = () => {
      if (isAppSettingScrollTop()) {
        // The app is setting the scrollTop.
        // Note: This doesn't always trigger on navigation events for some reason.
        setShouldHideHeader(false);
        return;
      }
      const { scrollTop } = scrollingElement;
      const dx = scrollTop - _prevScroll;
      _prevScroll = scrollTop;
      if (scrollTop === 0) {
        _shouldHideHeader = false;
        setShouldHideHeader(false);
      } else if (dx > 0) {
        // Scrolling down;
        if (scrollTop > headerPadding && !_shouldHideHeader) {
          _shouldHideHeader = true;
          setShouldHideHeader(true);
        }
      } else {
        // Scrolling up.
        if (
          // iPad registers scrolling when it drags past the end of the document.
          // Ensure the header doesn't come back when that happens.
          scrollingElement.scrollHeight - scrollTop > window.innerHeight &&
          _shouldHideHeader
        ) {
          _shouldHideHeader = false;
          setShouldHideHeader(false);
        }
      }
    };

    document.addEventListener('scroll', onScroll);
    return () => {
      document.removeEventListener('scroll', onScroll);
    };
  }, []);

  let isOpen;
  let title;
  switch (view) {
    case 'settings':
      isOpen = true;
      title = (
        <Path path="/" key={key} title="⚙️ Settings" hideSiteName={true} />
      );
      break;
    case 'file-storage':
      isOpen = true;
      title = (
        <Path
          path="/"
          key={key}
          title="Manage File Storage"
          hideSiteName={true}
        />
      );
      break;
    case 'privacy':
      isOpen = true;
      title = (
        <Path
          path="/"
          key={key}
          title="👀 Privacy Policy and Usage"
          hideSiteName={true}
        />
      );
      break;
    case 'view-file':
    case 'view-pdf':
    case 'view-image':
    case 'view-markdown':
    case 'language-coach':
      isOpen = false;
      title = <Path key={key} path={path} />;
      break;
    case 'list-files':
      if (location.pathname !== '/folder' && location.pathname !== '/') {
        isOpen = false;
        title = <Path key={key} path={path} />;
        break;
      }
    // fallthrough
    case null:
    default:
      isOpen = true;
      assertType<'list-files' | null>(view);
      title = (
        <div className="headerTitle" key={key}>
          <FileStoreSelection />
        </div>
      );
      break;
  }

  return (
    <div className="header" style={headerStyle}>
      <SiteName isOpen={isOpen} />
      <div className="headerStart">{title}</div>
      <div className="headerEnd">
        <SaveFileButton />
        <RequestFullScreen />
        <SettingsButton />
      </div>
    </div>
  );
}

function goFullScreen() {
  const element = document.querySelector('[data-fullscreen]');
  if (!element) {
    console.error('No [data-fullscreen] was found.');
    return;
  }
  if (element.requestFullscreen) {
    element.requestFullscreen().catch((error) => {
      console.error('Failed to go fullscreen', error);
    });
  }
  if ((element as any).webkitRequestFullscreen) {
    (element as any).webkitRequestFullscreen().catch((error: any) => {
      console.error('Failed to go fullscreen', error);
    });
  }
}

function fullScreenEventHandler(event: KeyboardEvent) {
  if (
    event.key === 'f' &&
    !event.shiftKey &&
    !event.ctrlKey &&
    !event.metaKey &&
    !event.altKey
  ) {
    goFullScreen();
  }
}

function RequestFullScreen() {
  const canGoFullScreen = $$.canGoFullScreen();
  React.useEffect(() => {
    if (canGoFullScreen) {
      document.addEventListener('keyup', fullScreenEventHandler);
    } else {
      document.removeEventListener('keyup', fullScreenEventHandler);
    }
    return () => {
      document.removeEventListener('keyup', fullScreenEventHandler);
    };
  }, [canGoFullScreen]);

  if (!canGoFullScreen) {
    return null;
  }

  return (
    <button type="button" className="button" onClick={goFullScreen}>
      Fullscreen
    </button>
  );
}

function SettingsButton() {
  const view = $$.getView();
  if (view === 'settings') {
    return null;
  }
  return (
    <a href="/settings" className="button headerSettings">
      Settings
    </a>
  );
}

function FileStoreSelection() {
  const fileStoreDisplayName = $$.getFileStoreDisplayName();
  const server = $$.getCurrentServerOrNull();
  const fileStoreServers = $$.getServers();
  const dispatch = Hooks.useDispatch();
  const button = React.useRef<null | HTMLButtonElement>(null);
  const [openEventDetail, setOpenEventDetail] = React.useState(-1);
  const [openGeneration, setOpenGeneration] = React.useState(0);
  const navigate = Router.useNavigate();
  const buttons = React.useMemo<MenuButton[]>(
    () => [
      {
        key: 'browser',
        onClick: () => {
          navigate('');
          dispatch(A.changeFileStore('browser'));
        },
        children: getBrowserName(),
      },
      {
        key: 'dropbox',
        onClick: () => {
          navigate('');
          dispatch(A.changeFileStore('dropbox'));
        },
        children: 'Dropbox',
      },
      ...fileStoreServers.map((server) => ({
        key: `fs-server-${server.url}-${server.name}`,
        onClick: () => {
          navigate('');
          dispatch(A.changeFileStore('server', server));
        },
        children: server.name,
      })),
      {
        key: 'add-fs-server',
        onClick: () => {
          navigate('/add-file-storage');
        },
        children: 'Add File Storage',
      },
    ],
    [],
  );

  return (
    <>
      <button
        type="button"
        className="headerFileStoreSelection"
        ref={button}
        title="Change the file system source"
        onClick={(event) => {
          setOpenGeneration((generation) => generation + 1);
          setOpenEventDetail(event.detail);
        }}
      >
        {fileStoreDisplayName}
      </button>
      {overlayPortal(
        <Menu
          clickedElement={button}
          openEventDetail={openEventDetail}
          openGeneration={openGeneration}
          buttons={buttons}
        />,
      )}
    </>
  );
}

function SaveFileButton() {
  const text = $$.getActiveFileTextOrNull();
  const dispatch = Hooks.useDispatch();
  const path = $$.getPath();
  const request = $$.getDownloadFileCache().get(path);
  const isModified = $$.getIsActiveFileModified();

  if (!isModified || text === null || !request) {
    return null;
  }

  return (
    <button
      className="button headerSaveFile"
      onClick={() => {
        dispatch(A.saveTextFile(path, text)).catch((error) => {
          console.error('Failed to save file', error);
        });
      }}
    >
      Save
    </button>
  );
}

function Path({
  path,
  title,
  hideSiteName,
}: {
  path: string;
  title?: any;
  hideSiteName?: boolean;
}) {
  const songTitle = $$.getActiveFileSongTitleOrNull();
  const fsSlug = $$.getCurrentFileStoreSlug();
  const breadcrumbs = [];
  let pathGrow = '';
  const parts = path.split('/');
  const fileName = parts.pop();
  for (const part of parts) {
    if (breadcrumbs.length === 0) {
      breadcrumbs.push(
        <Router.Link key="/" to="/">
          Home
        </Router.Link>,
      );
      continue;
    }

    pathGrow += '/' + part;

    if (part.endsWith('.coach')) {
      // Use special handling for language coach breadcrumbs.
      const section = $$.getLanguageCoachSection();
      breadcrumbs.push(
        <span key={pathGrow + '»'}>»</span>,
        <Router.Link
          key={pathGrow}
          to={`/${fsSlug}/language-coach${pathGrow}?section=${section}`}
        >
          {part}
        </Router.Link>,
      );
      break;
    }

    breadcrumbs.push(
      <span key={pathGrow + '»'}>»</span>,
      <Router.Link key={pathGrow} to={`/${fsSlug}/folder${pathGrow}`}>
        {part}
      </Router.Link>,
    );
  }
  const backParts = path.split('/');
  backParts.pop();

  return (
    <>
      <div className="headerPath headerPathFull" key={'full' + path}>
        <SiteName isOpen={!hideSiteName && !(songTitle ?? fileName)} />
        <FileStoreSelection key="fileStore" />
        <div className="headerPathBreadcrumbs">
          {breadcrumbs}
          <span>»</span>
          {songTitle ?? fileName ? (
            <span title={fileName}>{songTitle ?? fileName}</span>
          ) : null}
          {title ? <span>{title}</span> : null}
        </div>
      </div>
      <div className="headerPath headerPathMobile" key={'mobile' + path}>
        <Router.Link
          to={`/${fsSlug}/folder${backParts.join('/')}`}
          className="headerPathBack"
          aria-label="back"
        ></Router.Link>
      </div>
    </>
  );
}

interface SlideInProps {
  isOpen: boolean;
  skipAnimation?: boolean;
  children: any;
}

function SlideIn({ isOpen, skipAnimation, children }: SlideInProps) {
  const contentsRef = React.useRef<HTMLDivElement | null>(null);
  const [width, setWidth] = React.useState<number>(0);

  React.useEffect(() => {
    const { current } = contentsRef;
    if (!current) {
      return;
    }
    const rect = current.getBoundingClientRect();
    setWidth(rect.width);
  }, [contentsRef]);

  const className = skipAnimation
    ? 'headerSlideIn'
    : 'headerSlideIn headerSlideInAnimate';

  return (
    <div className={className} style={{ width: isOpen ? width : 0 }}>
      <div className="headerSlideInContents" ref={contentsRef}>
        {children}
      </div>
    </div>
  );
}

let wasAnimated = false;
function SiteName(props: { isOpen: boolean }) {
  const [skipAnimation, setSkipAnimation] = React.useState(!wasAnimated);

  React.useEffect(() => {
    if (wasAnimated) {
      setSkipAnimation(false);
    }
    wasAnimated = true;
  }, [props.isOpen]);

  if (process.env.SITE === 'floppydisk') {
    return (
      <div className="headerSiteName">
        <SlideIn isOpen={props.isOpen} skipAnimation={skipAnimation}>
          <span>
            <img src="/favicon-48x48.png" className="headerFloppyDiskImg" />{' '}
          </span>
          <span className="headerSiteNameTitle">
            FloppyDisk<span className="headerSiteNameSuffix">.link</span>
          </span>
        </SlideIn>
      </div>
    );
  } else {
    return (
      <div className="headerSiteName">
        <SlideIn isOpen={props.isOpen} skipAnimation={skipAnimation}>
          <div className="headerTitleSlideIn">
            <span>🎵 </span>
            <span className="headerSiteNameTitle">Browser Chords</span>
          </div>
        </SlideIn>
      </div>
    );
  }
}
