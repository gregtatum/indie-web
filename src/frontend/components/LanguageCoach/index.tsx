import * as React from 'react';
import { A, T, $$, Hooks } from 'frontend';
import { pathJoin, UnhandledCaseError } from 'frontend/utils';
import { Learned } from './Learned';
import { HomePage } from './HomePage';
import { StudyList } from './StudyList';
import { LCHeader } from './LCHeader';
import { Reading } from './Reading';
import './index.css';

export function LanguageCoach() {
  const lcPath = $$.getLanguageCoachPath();
  const dataOrNull = $$.getLanguageCoachDataOrNull();
  const fileSystem = $$.getCurrentFS();
  const fsName = $$.getCurrentFileSystemName();
  const dispatch = Hooks.useDispatch();

  React.useEffect(() => {
    if (!dataOrNull) {
      let cachedHash = '';
      let fileSystemTextReceived = false;

      // Attempt to pull it from the cache quickly.
      fileSystem.cache
        ?.loadText(pathJoin(lcPath, 'words.json'))
        .then((text) => {
          if (fileSystemTextReceived) {
            // The cache lost the race.
            return;
          }
          cachedHash = text.metadata.hash;
          console.log('Loading from the indexeddb');
          const data = JSON.parse(text.text);
          dispatch(A.loadLanguageData(data));
        })
        .catch(console.error);

      // Kick off the potentially slower request that can go over the network.
      fileSystem.loadText(pathJoin(lcPath, 'words.json')).then(
        (text) => {
          try {
            if (cachedHash === text.metadata.hash) {
              return;
            }
            fileSystemTextReceived = true;
            const data = JSON.parse(text.text);
            dispatch(A.loadLanguageData(data));
          } catch (error) {
            console.error(error);
          }
        },
        (error) => {
          console.error(error);
          dispatch(
            A.addMessage({
              message: <>Failed to load language data.</>,
              timeout: true,
            }),
          );
        },
      );
    }
  }, [dataOrNull]);

  if (!dataOrNull) {
    return <div className="centered">Loading Language Coach dataâ€¦</div>;
  }
  return (
    <DataSync key={lcPath} data={dataOrNull}>
      <div className="language-coach" key={fsName + lcPath}>
        <LCHeader />
        <Sections />
      </div>
    </DataSync>
  );
}

function Sections() {
  const section = $$.getLanguageCoachSection();
  switch (section) {
    case 'home':
      return <HomePage />;
    case 'study-list':
      return <StudyList />;
    case 'learned':
      return <Learned />;
    case 'reading':
      return <Reading />;
    default:
      throw new UnhandledCaseError(section, 'Unhandled view');
  }
}

/**
 * Syncs the data based off of a time out.
 */
function DataSync(props: {
  key: string;
  data: T.LanguagCoachDataState;
  children: any;
}) {
  const { data, children } = props;

  const dispatch = Hooks.useDispatch();

  // These can be used in the effect freely to manage state in the effect.
  const timeoutId = React.useRef<number>(-1);
  const unloadHandler = React.useRef<null | (() => string)>();
  const savePromise = React.useRef<Promise<unknown>>(Promise.resolve());
  const unloadMessageGeneration = React.useRef<null | number>(null);
  const firstLoad = React.useRef<boolean>(true);

  const fileSystem = $$.getCurrentFS();
  const fullPath = pathJoin($$.getLanguageCoachPath(), 'words.json');

  const { language, learnedStems, ignoredStems } = data;
  const invalidations = [
    language,
    learnedStems,
    ignoredStems,
    fileSystem,
    fullPath,
  ];

  React.useEffect(() => {
    if (firstLoad.current) {
      firstLoad.current = false;
      return;
    }
    clearTimeout(timeoutId.current);

    if (!unloadHandler.current) {
      unloadHandler.current = () => {
        unloadMessageGeneration.current = dispatch(
          A.addMessage({
            message: (
              <>
                Saving <code>{fullPath}</code>
              </>
            ),
          }),
        );
        return 'There is a pending language coach save, are you sure you want to leave?';
      };
      window.addEventListener('beforeunload', unloadHandler.current);
    }

    async function saveLanguageCoatchData() {
      try {
        // If there is a previous save attempt, wait for it to be settled.
        await savePromise.current.catch(() => {});

        const serializedData: T.LanguageDataV1 = {
          description: 'The data store for the language coach',
          lastSaved: Date.now(),
          language,
          version: 1,
          learnedStems: [...learnedStems],
          ignoredStems: [...ignoredStems],
        };

        // Kick off the save.
        savePromise.current = fileSystem.saveText(
          fullPath,
          'overwrite',
          JSON.stringify(serializedData, null, 2),
        );

        await savePromise.current;

        if (unloadMessageGeneration.current !== null) {
          // We finished saving after a user tried to leave the page.
          dispatch(
            A.addMessage({
              message: (
                <>
                  Finished saving <code>{fullPath}</code>. Feel free to close
                  the tab.
                </>
              ),
              generation: unloadMessageGeneration.current,
            }),
          );
        }
      } catch (error) {
        // The save failed, notify the user.
        console.error(error);
        dispatch(
          A.addMessage({
            message: (
              <>
                Unable to save <code>{fullPath}</code>
              </>
            ),
          }),
        );
      }

      if (unloadHandler.current) {
        window.removeEventListener('beforeunload', unloadHandler.current);
        unloadHandler.current = null;
      }
      unloadMessageGeneration.current = null;
    }

    timeoutId.current = setTimeout(
      saveLanguageCoatchData,
      5000,
    ) as unknown as number;
  }, invalidations);

  return children;
}
