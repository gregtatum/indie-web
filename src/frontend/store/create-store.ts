/**
 * This file is coercing things into the shape of the store, so be permissive about
 * `any` types here.
 */

import {
  legacy_createStore as reduxCreateStore,
  applyMiddleware,
  type Middleware,
} from 'redux';
import { thunk } from 'redux-thunk';
import { mainReducer } from 'frontend/store/reducers';
import { Store, Action, State } from 'frontend/@types';
import { T } from 'frontend';

/**
 * Create a more minimalist action logger.
 */
export const logger =
  (store: Store) => (next: (action: Action) => any) => (action: Action) => {
    const style = 'font-weight: bold; color: #fa0';
    const prevState = store.getState();
    const result = next(action);
    const nextState = store.getState();
    console.log(`[action] %c${action.type}`, style, {
      action,
      prevState,
      nextState,
      stack: (new Error().stack || '(no stack)').split('\n'),
    });
    return result;
  };

/**
 * Isolate the store creation into a function, so that it can be used outside of the
 * app's execution context, e.g. for testing.
 */
export function createStore(): Store {
  const middlewares: Middleware[] = [thunk];

  if (process.env.NODE_ENV !== 'test') {
    middlewares.push(logger as any);
  }
  const store = reduxCreateStore(
    mainReducer as any,
    applyMiddleware(...middlewares),
  );

  return store as any;
}

function blobToBase64(blob: Blob | undefined): Promise<undefined | string> {
  if (!blob) {
    return Promise.resolve(undefined);
  }
  return new Promise((resolve, reject) => {
    const reader = new FileReader();
    reader.readAsDataURL(blob);
    reader.onloadend = () => {
      const { result } = reader;
      if (typeof result === 'string') {
        resolve(result);
      } else {
        reject(new Error());
      }
    };
  });
}

export function base64toBlob(
  b64Data: string,
  contentType = '',
  sliceSize = 512,
): Blob {
  const byteCharacters = atob(b64Data);
  const byteArrays = [];

  for (let offset = 0; offset < byteCharacters.length; offset += sliceSize) {
    const slice = byteCharacters.slice(offset, offset + sliceSize);

    const byteNumbers = new Array(slice.length);
    for (let i = 0; i < slice.length; i++) {
      byteNumbers[i] = slice.charCodeAt(i);
    }

    const byteArray = new Uint8Array(byteNumbers);
    byteArrays.push(byteArray);
  }

  return new Blob(byteArrays, { type: contentType });
}

async function serializeDownloadBlobCache(
  downloadBlobCache: T.DownloadBlobCache,
): Promise<any> {
  const cache: Array<any> = [];
  for (const [k, { metadata, blob }] of downloadBlobCache.entries()) {
    cache.push([
      k,
      {
        metadata,
        blob: await blobToBase64(blob),
      },
    ]);
  }
  return cache;
}

// TODO - This could be written more type safe with a Entries<Map<K,V>> type
// and with a ToRecord<Interface> type. A pseudo-recursive Serializable type
// can be made with:
//
// type SerializableImpl<T> = Record<string, T> | null | string | number;
// type Serializable = SerializeImpl<SerializeImpl<SerializeImpl<never>>>;
export async function serializeState(state: State): Promise<unknown> {
  const { listFilesCache, downloadFileCache, downloadBlobCache } = state;

  return {
    ...state,
    listFilesCache: [...listFilesCache.entries()],
    downloadFileCache: [...downloadFileCache.entries()],
    downloadBlobCache: await serializeDownloadBlobCache(downloadBlobCache),
  };
}

export function deserializeState(state: any): State {
  const { listFilesCache, downloadFileCache, downloadBlobCache } = state.app;
  return {
    ...state,
    listFilesCache: new Map(listFilesCache),
    downloadFileCache: new Map(downloadFileCache),
    downloadBlobCache: new Map(downloadBlobCache),
  };
}
