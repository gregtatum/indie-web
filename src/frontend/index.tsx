import * as React from 'react';
import { Provider, useSelector } from 'react-redux';
import { createStore } from './store/create-store';
import * as T from 'frontend/@types';
import { App } from 'frontend/components/App';
import { ensureExists, maybeMockGoogleAnalytics } from 'frontend/utils';
import { createRoot } from 'react-dom/client';

import * as A from 'frontend/store/actions';
import * as $ from 'frontend/store/selectors/index';
import {
  BROWSER_FILES_DB_NAME,
  openIDBFS,
} from './logic/file-system/indexeddb-fs';

export * as A from 'frontend/store/actions';
export * as $ from 'frontend/store/selectors/index';
export * as T from 'frontend/@types';
export * as Hooks from 'frontend/hooks';

type HookedSelectors = {
  [FnName in keyof typeof $]: () => ReturnType<(typeof $)[FnName]>;
};

/**
 * Exports all of the selectors as hooks that can be used directly inside of
 * a React component.
 *
 * e.g. instead of:
 *
 * const view = Hooks.useSelector($.getView);
 *
 * Write:
 *
 * const view = $$.getView();
 */
export const $$: HookedSelectors = {} as any;
for (const [name, fn] of Object.entries($)) {
  ($$ as any)[name] = () => useSelector(fn as any);
}

if (process.env.NODE_ENV !== 'test') {
  document.addEventListener('DOMContentLoaded', () => {
    init().catch((error) => {
      console.error('Error during initialization', error);
    });
  });
}

export async function init(): Promise<void> {
  maybeMockGoogleAnalytics();
  initServiceWorker();

  const store = createStore();

  // Expose the $$ as a global for the webconsole.
  const $$ = {};
  for (const [name, fn] of Object.entries($)) {
    ($$ as any)[name] = () => (fn as any)(store.getState());
  }

  Object.assign(window as any, {
    store,
    dispatch: store.dispatch,
    getState: store.getState,
    $,
    $$,
    A,
    expireOauth() {
      const oauth = $.getDropboxOauth(store.getState());
      if (!oauth) {
        return;
      }
      store.dispatch(
        A.setDropboxAccessToken(oauth.accessToken, 0, oauth.refreshToken),
      );
    },
  });

  const cachePromise = $.getCurrentFSOrNull(store.getState())?.cachePromise;
  if (cachePromise) {
    await cachePromise;
  }
  const ipdbfs = await openIDBFS(BROWSER_FILES_DB_NAME);
  store.dispatch(A.connectIDBFS(ipdbfs));

  mountReact(store);
}

export function createRootApp(store: T.Store): JSX.Element {
  return (
    <Provider store={store as any}>
      <App key="app" />
    </Provider>
  );
}

function mountReact(store: T.Store): void {
  const mountElement = ensureExists(
    document.querySelector('.appRoot'),
    'Could not find the app root',
  );
  const root = createRoot(mountElement);
  root.render(createRootApp(store));
}

function initServiceWorker() {
  if (
    process.env.NODE_ENV === 'development' ||
    process.env.NODE_ENV === 'test'
  ) {
    return;
  }
  window.addEventListener('load', async () => {
    try {
      const registration =
        await navigator.serviceWorker.register('/service-worker.js');
      console.log('Service worker registered: ', registration);
    } catch (error) {
      console.log('Service worker registration failed:', error);
    }
  });
}
