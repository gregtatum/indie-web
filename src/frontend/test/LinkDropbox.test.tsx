import { createStore } from 'frontend/store/create-store';
import { App } from 'frontend/components/App';
import { render, screen, waitFor } from '@testing-library/react';
import * as React from 'react';
import { Provider } from 'react-redux';
import { A } from 'frontend';

describe('LinkDropbox', () => {
  it('can render', async () => {
    const store = createStore();
    store.dispatch(A.changeFileStore('dropbox'));
    const { container } = render(
      <Provider store={store as any}>
        <App />
      </Provider>,
    );

    await waitFor(() => screen.getByText(/Store Files on Dropbox/));

    const button = screen.getByText(/Connect Dropbox/);
    await waitFor(() => expect(button.getAttribute('href')).toBeTruthy());

    expect(container.firstChild).toMatchSnapshot();
  });
});
