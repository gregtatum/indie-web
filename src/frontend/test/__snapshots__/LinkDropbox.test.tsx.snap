// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LinkDropbox can render 1`] = `
<div
  class="appView"
>
  <div
    class="header"
  >
    <div
      class="headerSiteName"
    >
      <div
        class="headerSlideIn"
        style="width: 0px;"
      >
        <div
          class="headerSlideInContents"
        >
          <div
            class="headerTitleSlideIn"
          >
            <span>
              ðŸŽµ 
            </span>
            <span
              class="headerSiteNameTitle"
            >
              Browser Chords
            </span>
          </div>
        </div>
      </div>
    </div>
    <div
      class="headerStart"
    >
      <div
        class="headerTitle"
      >
        <button
          class="headerFileSystemSelection"
          title="Change the file system source"
          type="button"
        >
          Dropbox
        </button>
      </div>
    </div>
    <div
      class="headerEnd"
    >
      <a
        class="button headerSettings"
        href="/settings"
      >
        Settings
      </a>
    </div>
  </div>
  <div
    class="linkDropbox"
  >
    <div
      class="linkDropboxContent"
    >
      <div
        class="linkDropboxDescription"
      >
        <h1
          class="linkDropboxH1"
        >
          Store Files on Dropbox
        </h1>
        <h2
          class="linkDropboxH2"
        >
          Take your files across devices
        </h2>
        <p>
          Manage tabs, chords, and sheet music in Dropbox. Access and edit them directly in the browser, from anywhere.
        </p>
        <p>
          Privacy is important. 
          Browser Chords
           will only be given access to the 
          <code>
            Dropbox/Apps/Chords
          </code>
           
          folder in Dropbox to manage files. See the
           
          <a
            href="/privacy"
          >
            privacy policy
          </a>
           for more details. The source code is on
           
          <a
            href="https://github.com/gregtatum/indie-web"
          >
            GitHub
          </a>
          .
        </p>
      </div>
      <div>
        <a
          class="linkDropboxConnect"
          href="https://www.dropbox.com/oauth2/authorize?response_type=code&code_challenge_method=S256&client_id=FAKE_DROPBOX_CLIENT_ID&code_challenge=fakebase64&redirect_uri=http%3A%2F%2Flocalhost%2Flogin&token_access_type=offline"
        >
          Connect Dropbox
        </a>
      </div>
    </div>
  </div>
</div>
`;
